### PR TITLE
log: Add per target log levels

### PR DESCRIPTION
**Pull request by daandemeyer**
For CI in mkosi, I want to configure systemd to log at debug level
to the journal, but not to the console. While we already have max
level settings for journald's forwarding settings, not every log line
goes to the journal, specifically during early boot and when units
are connected directly to the console (think systemd-firstboot), so
let's extend the log level options we already have to allow specifying
a comma separated list of values and lets allow prefixing values with
the log target they apply to to make this possible.

### CODE REVIEW

**Review by bluca — APPROVED** [review]

**Review by yuwata — COMMENTED** [review]
Hmm, how about to make `$SYSTEMD_LOG_LEVEL`, `--log-level=` and friends take extended string? E.g. `SYSTEMD_LOG_LEVEL=debug,console:info,syslog:info,kmsg:critical`.

**Comment by DaanDeMeyer** [comment]
That's a great idea! Will rework it like that.

**Comment by YHNdnzj** [comment]
> That's a great idea! Except that it will make integration in mkosi harder since the new format won't be understood on distributions with older systemd.

Hmm, it is possible to have two kernel cmdline options with old and new format at the same time, and older systemd versions will ignore the new one? (Although it's definitely a dirty workaround...)

**Comment by DaanDeMeyer** [comment]
Updated to the suggested approach

**Comment by DaanDeMeyer** [comment]
Putting this as rc-blocker as it's going to be extremely useful for CI

File: src/basic/log.c [hunk 1]
```diff
@@ -1215,14 +1233,87 @@ int log_set_target_from_string(const char *e) {
...
+                r = log_level_to_string_alloc(log_target_max_level[target], &l);
+                if (r < 0)
+                        return r;
+
+                q = strjoin(log_target_to_string(target), ":", l);
+                if (!q)
+                        return -ENOMEM;
+
+                if (!strextend_with_separator(&s, ",", q))
+                        return -ENOMEM;
```
Comment (YHNdnzj) [inline src/basic/log.c]: strextendf_with_separator

File: man/common-variables.xml [hunk 2]
```diff
@@ -11,12 +11,15 @@
...
+      log level, i.e. less important ones, will be suppressed). Takes a comma-separated list of values. A
+      value may be either one of (in order of decreasing importance) <constant>emerg</constant>,
+      <constant>alert</constant>, <constant>crit</constant>, <constant>err</constant>,
+      <constant>warning</constant>, <constant>notice</constant>, <constant>info</constant>,
+      <constant>debug</constant>, or an integer in the range 0…7. See
       <citerefentry project='man-pages'><refentrytitle>syslog</refentrytitle><manvolnum>3</manvolnum></citerefentry>
-      for more information.</para>
+      for more information. Each value may optionally be prefixed with one of <constant>console</constant>,
+      <constant>syslog</constant>, <constant>kmsg</constant> or <constant>journal</constant> followed by a
+      colon to set the maximum log level for that specific log target.</para>
```
Comment (yuwata) [inline man/common-variables.xml]: Maybe, better to add an example of the extended format?

File: src/basic/log.c [hunk 3]
```diff
@@ -439,6 +445,9 @@ static int write_to_console(
         if (console_fd < 0)
                 return 0;
 
+        if (_likely_(LOG_PRI(level) > log_target_max_level[LOG_TARGET_CONSOLE]))
```
Comment (yuwata) [inline src/basic/log.c]: The per target max level defaults to `INT_MAX`, so is this condition `_likely_` satisfied??

File: src/basic/log.c [hunk 4]
```diff
@@ -49,6 +49,12 @@ static void *log_syntax_callback_userdata = NULL;
 
 static LogTarget log_target = LOG_TARGET_CONSOLE;
 static int log_max_level = LOG_INFO;
+static int log_target_max_level[] = {
+        [LOG_TARGET_CONSOLE] = INT_MAX,
+        [LOG_TARGET_KMSG]    = INT_MAX,
+        [LOG_TARGET_SYSLOG]  = INT_MAX,
+        [LOG_TARGET_JOURNAL] = INT_MAX,
```
Comment (yuwata) [inline src/basic/log.c:56]: Maybe, better to reorder the definition of LOG_TARGET_XYZ, and define these four earlier, and add _LOG_TARGET_SINGLE_FUNC or so to distinguish them from others like _JOURNAL_OR_KMSG ?

File: src/basic/log.c [hunk 5]
```diff
@@ -1215,14 +1233,84 @@ int log_set_target_from_string(const char *e) {
         return 0;
 }
 
+static bool log_target_has_max_log_level(LogTarget target) {
+        return IN_SET(target, LOG_TARGET_KMSG, LOG_TARGET_JOURNAL, LOG_TARGET_SYSLOG, LOG_TARGET_CONSOLE);
```
Comment (yuwata) [inline src/basic/log.c]: Then, we can make it much simpler?

**Review by yuwata — APPROVED** [review]
LGTM. Several minor comments. Feel free to ignore them.

**Comment by DaanDeMeyer** [comment]
I think I made sufficient changes to require another review, PTAL!

File: man/common-variables.xml [hunk 6]
```diff
@@ -11,13 +11,17 @@
...
+      <constant>alert</constant>, <constant>crit</constant>, <constant>err</constant>,
+      <constant>warning</constant>, <constant>notice</constant>, <constant>info</constant>,
+      <constant>debug</constant>, or an integer in the range 0…7. See
       <citerefentry project='man-pages'><refentrytitle>syslog</refentrytitle><manvolnum>3</manvolnum></citerefentry>
-      for more information.</para>
-      </listitem>
+      for more information. Each value may optionally be prefixed with one of <constant>console</constant>,
+      <constant>syslog</constant>, <constant>kmsg</constant> or <constant>journal</constant> followed by a
+      colon to set the maximum log level for that specific log target (e.g.
+      <constant>SYSTEMD_LOG_LEVEL=debug,console:info</constant> specifies to log at debug level except when
```
Comment (YHNdnzj) [inline man/common-variables.xml:23]: Should probably be specific that per-target settings will not be able to go above the global one?

File: src/basic/log.c [hunk 7]
```diff
@@ -1218,11 +1236,74 @@ int log_set_target_from_string(const char *e) {
...
+                if (!colon) {
+                        r = log_level_from_string(word);
+                        if (r < 0)
+                                return r;
+
+                        log_set_max_level(r);
+                        continue;
+                }
+
+                prefix = strndup(word, colon - word);
```
Comment (YHNdnzj) [inline src/basic/log.c:1260]: Maybe just `strndupa_safe`?

File: man/udevadm.xml [hunk 8]
```diff
@@ -662,13 +662,10 @@
...
           <listitem>
             <para>Set the internal log level of
-            <filename>systemd-udevd</filename>. Valid values are the
-            numerical syslog priorities or their textual
-            representations: <option>emerg</option>,
-            <option>alert</option>, <option>crit</option>,
-            <option>err</option>, <option>warning</option>,
-            <option>notice</option>, <option>info</option>, and
-            <option>debug</option>.</para>
+            <filename>systemd-udevd</filename>. Accepts the same values as <option>--log-level=</option>
```
Comment (YHNdnzj) [inline man/udevadm.xml]: Hmm, most of our code (including udev) still uses `log_level_from_string`? So I don't see how this would work?

**Review by YHNdnzj — APPROVED** [review]
Looks good, but I don't see how this would work for udev. Maybe drop that part and we can teach `config_parse_log_level` and friends to honor the new format in the future?

File: src/basic/log.c (hunk 7 above)
Comment (daandemeyer) [reply src/basic/log.c:1260]: That's just premature optimization if you ask me

File: man/udevadm.xml (hunk 8 above)
Comment (daandemeyer) [reply man/udevadm.xml]: Dropped, not sure how I thought this could work.

File: man/common-variables.xml [hunk 9]
```diff
@@ -11,13 +11,18 @@
...
+      logging to the console which should be at info level). Note that the global maximum log level takes
+      priority over any per target maximum log levels.</para></listitem>
```
Comment (poettering) [inline man/common-variables.xml:25]: hmm, we keep the overall log level? why make this two-level? why isn't the per-target log level enough? I think it is quite confusing to have two levels in place, that both filter. 

might be fine to automatically distill an overall-level from the per-target level (i.e. take the max of them), but i don't really see the point in keep the overall-level distinct from the per-target level.

**Comment by poettering** [comment]
so previously, we had one level and one target selector. With this we now have per-target levels. Which makes me wonder why we still need the target selector.

or in other words, wouldn't it make sense to fold $SYSTEMD_LOG_TARGET into this, so that its effect is to just adjusts the levels of the select target?

there was a request somewhere for us to allow logging to multiple places simulatneously. if you add the per-target levels anyway, then maybe make it work like that?

**Pull request merged**